com.unity.multiplayer.mlapi/Tests/Runtime/Messaging/NamedMessageTests.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Collections;
+using System.IO;
+using System.Text;
+using NUnit.Framework;
+using UnityEngine;
+using UnityEngine.TestTools;
+
+namespace MLAPI.RuntimeTests.Messaging
+{
+    public class NamedMessageTests : BaseMultiInstanceTest
+    {
+        protected override int NbClients => 1;
+
+        private NetworkManager Client => m_ClientNetworkManagers[0];
+
+        [UnityTest]
+        public IEnumerator NamedMessageIsReceivedOnClientWithContent()
+        {
+            var messageName = Guid.NewGuid().ToString();
+            var messageContent = Guid.NewGuid().ToString();
+            using var messageStream = new MemoryStream(Encoding.UTF8.GetBytes(messageContent));
+
+            m_ServerNetworkManager.CustomMessagingManager.SendNamedMessage(messageName, Client.LocalClientId, messageStream);
+
+            ulong receivedMessageSender = 0;
+            string receivedMessageContent = null;
+            m_ClientNetworkManagers[0].CustomMessagingManager.RegisterNamedMessageHandler(
+                messageName,
+                (sender, stream) =>
+                {
+                    receivedMessageSender = sender;
+
+                    using var memoryStream = new MemoryStream();
+                    stream.CopyTo(memoryStream);
+                    receivedMessageContent = Encoding.UTF8.GetString(memoryStream.ToArray());
+                });
+
+            yield return new WaitForSeconds(0.2f);
+
+            Assert.NotNull(receivedMessageContent);
+            Assert.AreEqual(messageContent, receivedMessageContent);
```

Comment:
Expected string length 36 but was 251. Strings differ at index 36.
  Expected: "...32-7101-4e3e-8197-214df3b7bc4c"
  But was:  "...32-7101-4e3e-8197-214df3b7bc4c4c\0\0\0\0\0\0\0�\0\0\0�\0\0\0\0\0\0\0\0\0..."

@ShadauxCat still happens after merging develop which contained the fixes you mentioned. There seems to be additional bytes as well, it was just straight nulls previously.

Reply:
So I've spent some time investigating this. This is not an issue with NamedMessages, but an issue with how the test is written (sort of) - which is to say, this test would have failed before the message ordering refactor, because NetworkBuffer.CopyTo() doesn't copy just up to the written length. It copies the entire buffer up to capacity. The NetworkBuffer's allocated with a capacity of 256, so even though `stream.Length` in your test is only 41, it's copying 251 bytes and that's adding junk to the end.

The difference between straight nulls vs random bytes now is just luck. It's junk memory. It could contain literally anything.

Reply:
That's interesting.

As a user, that's not how I would expect my payload to be handled though.

Reply:
Oh, I agree, but it's a bug in NetworkBuffer and it predates any of the changes I've recently made, and I don't really fully understand whether or not there's a reason it works that way, so I'm hesitant to change it.

Reply:
Okay yeah that makes sense. Anyone we can tag who can help out with this?

Reply:
Looks like this code came from outside of Unity back in 2018... so...

On the one hand I'd say let's just fix it... on the other hand, this class is going away soon, so should we bother?

Reply:
Actually, Git log tells me @TwoTenPvP is the chosen one.

Reply:
@ShadauxCat I'd be fine with just changing the assertion in the test to .StartsWith for this to pass and we can at least get the payload across the network, even if it contains trailing junk, which is better than no payload at all. Assuming that this gets logged as a bug and handled with hopefully soon.

Reply:
I pushed a change to NetworkBuffer. I can't see any reason we'd want it copying past length regardless.